README.md
# react-animated-progress-bar

Animated progress bar using react

### Use case scenario
![Circular bar](./images/usecase.png)

## Description

This package allows you to customize both circular and rectangular progress bar to various colors and sizes that suits your application or design.

- Progress bar animates when component is in view. By default it uses the document reference to check if component is in view or not. But you can pass the ref object which is scrollable to the scollArea props.

**For example**

````
const myRef = document.querySelector('.scrollable-div')

<ProgressBar width="230" trackWidth="13" percentage={score} scrollArea={myRef} />
````

**Otherwise**
if scrollArea is not defined, document reference object is used.

## Set up

To use package, Start by installing package

- npm i react-animated-progress-bar

**on your react project file**

### For a circular progress bar

```
import ProgressBar from 'react-animated-progress-bar';

<ProgressBar width="230" trackWidth="13" percentage="60" />
```

#### Output

![Circular bar](./images/60percentImg.png)

### For a rectangular progress bar

Set the rect props on the component to true

```
import ProgressBar from 'react-animated-progress-bar';

<ProgressBar
        width="400px"
        height="10px"
        rect
        fontColor="gray"
        percentage="70"
        rectPadding="1px"
        rectBorderRadius="20px"
        trackPathColor="transparent"
        bgColor="#333333"
        trackBorderColor="grey"
      />
```

#### Output

![Rect bar](./images/rectbar.png)

## Properties and description

- width (must be a string): width of the progressbar in rectangle type and diameter for circular type,

- percentage (must be a string): percentage of the progress bar

- height (must be a string): height for rectangular bar. Needed only in rectangular type

- rectPadding (must be a string): Padding around rectangular bar,

- trackBorderColor (must be a string): Border color of the track

- fontColor (must be a string): Percent font color

- rectBorderRadius (must be a string): border radius of rectangular bar

- trackPathColor (must be a string): color of the track path

- hollowBackgroundColor (must be a string): required only in circular bar. it gives the background color of the circle

- scrollArea (must be a dom element): the scrollable area. By default it is document

- defColor (must be an object); defines custom color for each level of the percentage ( poor, fair, good, excellent)

Score | Level | Color
----- | ----- |------
poor | 0 - 25 | #F32013
fair | 25 - 50 | #FF6700
good | 50 - 70 | #FFD900
excellent | 70 - 100 | #48AE2C


### Use case for the props in circular mode
````
const myRef = document.querySelector('.scrollable-div')

<ProgressBar
          width="500px"
          height="10px"
          fontColor="white"
          trackWidth="10"
          percentage="40"
          trackPathColor="grey"
          trackBorderColor="black"
          hollowBackgroundColor="#333333"
          defColor={{
            fair: 'orangered',
            good: 'yellow',
            excellent: 'green',
            poor: 'red',
          }}
          scrollArea={myAppRef}
        />
````

### output
![Circular bar](./images/darkbar.png)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)

### Fixed

## [1.1.1] - 2017-03-14

- PATCH
  Remove eventlistener from scroll after animation has been triggered

### Fixed

## [1.1.0] - 2017-03-14

- MINOR [ISS-9](https://github.com/kevoese/react-animated-progress-bar/issues/9)
  Fix issue with progress bar not updating


src/index.js
import React, { Component, createRef } from 'react';
import './app.css';
import {
  delay,
  contentInView,
  getPercentLevel,
  getRadius,
  propTypeValidation,
  makeCancelable,
  sanitizePercentage,
} from './utils';
import CircularProgressBar from './Components/CircularProgressBar';
import RectProgressBar from './Components/RectProgressBar';
class ProgressBar extends Component {
  state = {
    animate: false,
    counter: 0,
    responsiveRadius: 0,
    scrollAreaIsSet: null,
    stepDelay: null,
    countDelay: null,
  };

  myRef = createRef();
  rectTrackRef = createRef();
  trackRef = createRef();
  scrollRef = createRef();

  animateCount = async () => {
    try {
      this.setState({
        counter: 0,
      });
      const percentage = sanitizePercentage(this.props.percentage);
      const stepDelay = new delay(500);
      const countDelay = new delay(1000 / percentage);
      this.setState({
        stepDelay,
        countDelay,
      });
      await stepDelay.getPromise();

      while (this.state.counter < percentage) {
        await countDelay.getPromise();
        this.setState(prevState => ({
          ...prevState,
          counter: prevState.counter + 1,
        }));
      }
    } catch (error) {}
  };

  animateOnScroll = (shouldUpdate = false) => {
    if ((!this.state.animate && contentInView(this.myRef.current)) || shouldUpdate) {
      this.scrollRef && this.scrollRef.current && this.scrollRef.current.removeEventListener(
        'scroll',
        this.animateOnScroll
      );
      this.animateCount();
      this.setState({
        animate: true,
      });
      !this.props.rect
        ? this.trackRef.current.style.setProperty(
            '--level',
            getPercentLevel(this.props.percentage)
          )
        : this.rectTrackRef.current.style.setProperty(
            '--rectLevel',
            `${sanitizePercentage(this.props.percentage)}%`
          );
    }
  };

  componentDidMount() {
    this.animateOnScroll();
    this.setState({ scrollAreaIsSet: false });
  }

  componentDidUpdate(prevProps) {
    if (!this.state.scrollAreaIsSet) {
      this.setState({ scrollAreaIsSet: true });
      (this.props.scrollArea && typeof this.props.scrollArea == 'object')
        ? this.props.scrollArea.addEventListener('scroll', () => this.animateOnScroll(true))
        : document.addEventListener('scroll', () => this.animateOnScroll(true));
    }
    if(prevProps.percentage !== this.props.percentage) {
      this.animateOnScroll(true);
    }
  }

  componentWillUnmount() {
    // unsubscribe from timeouts and delay
    this.scrollRef && this.scrollRef.current && this.scrollRef.current.removeEventListener('scroll', this.animateOnScroll);
    this.state.stepDelay && this.state.stepDelay.cancel();
    this.state.countDelay && this.state.countDelay.cancel();
  }

  render() {
    const { width, rect } = this.props;
    return (
      <div
        ref={this.myRef}
        className="progress-bar"
        style={{ width: `${width}` }}
      >
        {!rect ? (
          <CircularProgressBar
            {...this.props}
            radius={getRadius(this.props.width)}
            animate={this.state.animate}
            counter={this.state.counter}
            trackRef={this.trackRef}
          />
        ) : (
          <RectProgressBar
            {...this.props}
            rectTrackRef={this.rectTrackRef}
            counter={this.state.counter}
            animate={this.state.animate}
          />
        )}
      </div>
    );
  }
}

ProgressBar.propTypes = propTypeValidation;

export default ProgressBar;


webpack.config.js
var path = require("path");
const MiniCssExtractPlugin = require("mini-css-extract-plugin");

module.exports = {
  mode: "production",
  entry: "./src/index.js",
  output: {
    path: path.resolve(__dirname, "dist"),
    filename: "index.js",
    libraryTarget: "commonjs2" // THIS IS THE MOST IMPORTANT LINE! :mindblow: I wasted more than 2 days until realize this was the line most important in all this guide.
  },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: /node_modules/,
        use: ["babel-loader"]
      },
      {
        test: /\.css$/,
        include: path.resolve(__dirname, "src"),
        exclude: /node_modules/,
        use: ["style-loader", "css-loader"]
      },
      {
        test: /\.(jpe?g|png|gif|mp3|svg|ico)$/,
        use: {
          loader: "file-loader",
          options: {}
        }
      }
    ]
  },
  plugins: [
    new MiniCssExtractPlugin({
      filename: "style.css"
    })
  ],
  externals: {
    react: {
      commonjs: "react",
      commonjs2: "react",
      amd: "React",
      root: "React"
    },
    "react-dom": {
      commonjs: "react-dom",
      commonjs2: "react-dom",
      amd: "ReactDOM",
      root: "ReactDOM"
    } // this line is just to use the React dependency of our parent-testing-project instead of using our own React.
  }
};


src/utils.js
export class delay {
  constructor(time) {
    this.time = time;
    this.timeout = null;
    this.close = null;
  }

  getPromise() {
    return new Promise((resolve, reject) => {
      this.close = reject;
      this.timeout = setTimeout(() => {
        resolve();
      }, this.time);
    });
  }
  cancel() {
    this.timeout && clearTimeout(this.timeout);
    this.close && this.close();
    return { isCanceled: true };
  }
}

export const contentInView = element => {
  const scroll = window.scrollY || window.pageYOffset;
  const elementPositionProps = element.getBoundingClientRect();
  const elementTopPosition = elementPositionProps.top + scroll;

  const viewport = {
    top: scroll,
    bottom: scroll + window.innerHeight,
  };

  const elementPosition = {
    top: Math.floor(elementTopPosition),
    bottom: Math.floor(elementTopPosition + elementPositionProps.height),
  };

  return (
    (elementPosition.bottom >= viewport.top &&
      elementPosition.bottom <= viewport.bottom) ||
    (elementPosition.top <= viewport.bottom &&
      elementPosition.top >= viewport.top)
  );
};

export const sanitizePercentage = percent => {
  let percentage = Number(percent);
  if (percentage < 0) percentage = 0;
  if (percentage > 100) percentage = 100;

  return percentage;
};

export const getPercentLevel = percentage => {
  let level = 252 - sanitizePercentage(percentage) * 2.52;
  return `${level}%`;
};

export const getColorType = (percentage, defColor) => {
  const percent = sanitizePercentage(percentage);
  const poor = (defColor && defColor.poor) || '#F32013';
  const fair = (defColor && defColor.fair) || '#ff6700';
  const good = (defColor && defColor.good) || 'rgb(255, 217, 0)';
  const excellent = (defColor && defColor.excellent) || '#48AE2C';

  const res =
    (percent < 25 && poor) ||
    (percent < 50 && fair) ||
    (percent < 70 && good) ||
    excellent;
  return res;
};

export const getNumber = str => {
  const arr = str.split('');
  let res = '';
  for (let i = 0; i < arr.length; i++) {
    if (isNaN(Number(arr[i]))) break;
    res += arr[i];
  }
  return res;
};

export const getRadius = raw => {
  const width = Number(getNumber(raw));
  if (width < 20 || isNaN(width)) return 250;
  const radius = raw.trim().slice(-1) === '%' ? width / 2 + '%' : width / 2;
  return radius;
};

export const propTypeValidation = {
  width: (props, propName) => {
    if (!props[propName] || typeof props[propName] != 'string')
      return new Error(
        `Invalid ${propName} supplied to the progress bar component. Please provide a width!`
      );
  },
  percentage: (props, propName) => {
    if (
      !props[propName] ||
      (props[propName] && typeof props[propName] != 'string')
    )
      return new Error(
        `Invalid ${propName} supplied to progress bar component. Please provide a percentage value!`
      );
  },
  height: (props, propName) => {
    if (
      props['rect'] &&
      (!props[propName] ||
        (props[propName] && typeof props[propName] != 'string'))
    )
      return new Error(
        `Invalid ${propName} supplied to progress bar component. Please provide a height!`
      );
  },
  rectPadding: (props, propName) => {
    if (
      props['rect'] &&
      (!props['height'] || typeof props['height'] != 'string')
    )
      return new Error(
        `Invalid information supplied to progress bar component. Please provide a height before you provide this ${propName}!`
      );
  },
  trackBorderColor: (props, propName) => {
    if (props[propName] && typeof props[propName] != 'string')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
  fontColor: (props, propName) => {
    if (props[propName] && typeof props[propName] != 'string')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
  rectBorderRadius: (props, propName) => {
    if (props[propName] && typeof props[propName] != 'string')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
  trackPathColor: (props, propName) => {
    if (props[propName] && typeof props[propName] != 'string')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
  hollowBackgroundColor: (props, propName) => {
    if (props[propName] && typeof props[propName] != 'string')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
  scrollArea: (props, propName) => {
    if (props[propName] && typeof props[propName] != 'object')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
  defColor: (props, propName) => {
    if (props[propName] && typeof props[propName] == 'object') {
      const arr = Object.keys(props[propName]);
      for (let i = 0; i < arr.length; i++) {
        if (typeof props[propName][arr[i]] != 'string')
          return new Error(
            `Invalid  property Key ${arr[i]} for ${propName} supplied to progress bar component!`
          );
      }
    } else if (props[propName] && typeof props[propName] !== 'object')
      return new Error(
        `Invalid ${propName} supplied to progress bar component!`
      );
  },
};
